File: basic/permutation/Permutation.py
import sys
sys.setrecursionlimit(10**6)

# Read input values for N and M
N, M = map(int, sys.stdin.readline().split())

sequence = [0] * M        # List to store the current sequence
visited = [False] * (N+1) # List to track if a number is used (1-indexed)
result_lines = []         # List to store the resulting sequences as strings

def dfs(depth):
    if depth == M:
        # Append the current sequence as a space-separated string
        result_lines.append(" ".join(map(str, sequence)))
        return
    for i in range(1, N+1):
        if not visited[i]:
            visited[i] = True
            sequence[depth] = i
            dfs(depth + 1)
            visited[i] = False  # Backtracking

dfs(0)
# Print all the results at once
sys.stdout.write("\n".join(result_lines))

File: basic/permutation/Permutation.js
// 백준
const fileAddr = '/dev/stdin'
// 로컬
//const fileAddr = __dirname + '/input.txt'

const input = require('fs')
  .readFileSync(fileAddr)
  .toString()
  .split('\n')
  .slice(0, -1);

const NM = input[0].split(" ");
const [n, m] = NM.map(Number);

function solution(n, m) {
    // Array to store the current sequence (length m)
    const seq = Array(m).fill(0);
    // Boolean array to track used numbers (1-indexed, so length n+1)
    const visited = Array(n + 1).fill(false);
    let result = "";

    // DFS function to generate all permutations
    function dfs(k) {
        if (k === m) {
            result += seq.join(' ') + "\n";
            return;
        }
        for (let i = 1; i <= n; i++) {
            if (!visited[i]) {
                seq[k] = i;
                visited[i] = true;
                dfs(k + 1);
                visited[i] = false;  // Backtracking
            }
        }
    }

    dfs(0);
    return result;
}

console.log(solution(n, m));
